pfc.8
.\" filter command-line reference
.TH PFC 8

.SH NAME
pfc \- A packet filter compiler

.SH SYNOPSIS
\fBpfc\fR [ \fB-c\fR | \fB--compile\fR ] [ \fB-t \fItarget\fB\fR | \fB--target=\fItarget\fB\fR ] [ \fB-o \fIoutfile\fB\fR | \fB--output=\fIoutfile\fB\fR ] \fIinfile\fR

\fBpfc\fR [ \fB-c\fR | \fB--compile\fR ] [ \fB-t \fItarget\fB\fR | \fB--target=\fItarget\fB\fR ] [ \fB-o \fIoutfile\fB\fR | \fB--output=\fIoutfile\fB\fR ] [ \fB-F \fIpolicy\fB\fR | \fB--flush=\fIpolicy\fB\fR ]

.SH DESCRIPTION
.B pfc
compiles packet filtering rules into a variety of formats from a
high-level description language.

.SH USAGE
.B pfc
reads the ruleset from the \fIinfile\fR specified on the command-line (or
standard input if \fIinfile\fR is "\-") and outputs to standard
output (or \fIoutfile\fR) via an optionally specified backend.

.PP
Both short and GNU-style long option options are accepted:

.TP
\fB-c, --compile\fR
Only try to "compile" the input, and do not generate any output.  This may be
useful to check that an input file has no syntax errors in it before one
attempts to use the result on a live server.

.TP
\fB-t \fItarget-filter\fB, --target=\fItarget-filter\fB\fR
If specified, \fItarget-filter\fR will be used to select an output filter type,
otherwise the default of \fBiptables\fR will be used.  Supported backends
are \fBiptables\fR, \fBipchains\fR, \fBipfilter\fR and \fBcisco\fR (for
Cisco IOS access-lists).

.TP
\fB-F \fIpolicy\fB, --flush=\fIpolicy\fB\fR
Flush mode.  Generate a set of rules for clearing all rules from the packet
filter.  Useful for firewall scripts that need to `shutdown' the firewall.
You can supply a \fIpolicy\fR argument in place of the usual filename, to
specify whether the flushed filter should default to \fBaccept\fR,
\fBreject\fR, or \fBdrop\fR.  It defaults to \fBaccept\fR, equivalent to
having no filter loaded at all.  It is not necessary to specify an
\fIinfile\fR when using flush mode.

.TP
\fB-o \fIoutfile\fB, --output=\fIoutfile\fB\fR
Write output to \fIoutfile\fR instead of standard output.

.SH BUGS
Not all backends implement all features.  

.SH SEE ALSO
\fBfilter_syntax\fR(5), \fBfilter_backends\fR(7)


HISTORY
0.13:
	Changed build system to SCons.
	Reorganised code structure.
	Added filtergen syntax emitter.
	Added option to turn off the resolver (useful for debugging!)
	Fixed bug with DROP targets at the end of forwaring chains.

0.12.8:
    Fix some easy to cause crashes.
    Remove some obsolete code that caused long compilation times.
    Prevents inclusion of the root filesystem path directly.
    Test that the example programs compile.

0.12.7:
	Supports iptables-restore output format.
	Supports IPv6 names.
	Supports ip6tables and ip6tables-restore output format.
	Better warnings when services don't resolve.
	Increases include file recursion depth to 512 from 16.
	Directory includes now have stable ordering of the files within.
	
	Runs tests in parallel so their output is captured, and generate more output.
	Adds more compiler warnings.
	Compiles clean under clang.

0.12.6:
	Support shell globbing in include directives.
	Allow '*' as a network interface name.

	Modernise build environment:
	Fix many compiler warnings.
	Finish migration to Git as upstream host.

0.12.5:
	Small fixes to test correctness.

0.12.4:
	Added directory include support.
	Fixed relative-path include error in fgadm.
	Improved tests for include file parsing.

0.12.3:
	Fix critical resolver bug.

0.12.2:
	New resolver code removes need for struct casting
	Altered test suite to be more robust
	minor cosmetic bugs

0.12.1:
	fixed some dumb bugs
	fixed some compile bugs on 64 bit platforms

0.12:
	converted build system to automake
	rewrote parser
	added test suite
	add "oneway" option
	fix iptables log message format (append a space)
	add "fgadm" command
	multi-homed host support

0.11:
	fixed non-working example in filter_syntax man page
	fix 64-bit warning in filter.c netmask calculation
	add "-F [policy]" flush option
	better feedback on parse errors

0.10:
	some cleanups, slight memory usage reduction
	add log messages: 'log text "foo etc"'
	fixed iptables and ipchains brace-grouping
	fix filter_unroll loop-making bug
	add RPM spec file from Wil Cooley <wcooley@nakedape.cc>
	rename package to "filtergen"
	add init scripts and makefile
	add samples for router, host and proxy firewall

0.9:
	fix install target to install section 7 manpages too
	do protocol lookups properly, support protocol numbers
	lookup/normalise service names
	force normalisation of netmasks
	add "-m" to force service and host names and to be
		turned into port numbers and IP addresses

0.8:
	cleanup for iptables and ipchains to save reallocs
	fix input vs output thing in forwarding support for iptables
		and ipchains

0.7:
	add "local" and "forward" syntax to state that packets are only
		for local or remote consumption (and "-h" option to
		default rules to the former)
	begin manpages
	add "-o" option to specify output file
	fix "make install"
	getopt()ify filtergen.c
	add "-l" option to minimise the number of rules for established
		connections, implemented for iptables backend
	add more stuff to the iptables skeleton
	warning fixups
	implement "local" and "forward" for iptables backend
	add "-r" option to mirror "-h"
	cleanup ipchains driver somewhat
	implement "local" for ipchains backend ("forward" is not possible)
	fix bug where accept would get lost from:
		input eth0 { ! dest bar drop; accept; };
	iptables and ipchains add "established packet" rules to head of chain
	add "proxy" as an alias for "redirect"

0.6:
	disable negation of conjunctions ("or"s, aka brace-grouping) for now
	add "no skeleton" option
	fix a couple of segfaults/aborts where input filter is invalid
	allow literal +, ", and _ signs in input
	add sub-groups -- where supported, large sub-sections can be moved
		out-of-line to improve performance

0.5: that's better
	much fixing of iptables driver.  Now shouldn't leak
	like the proverbial sieve

0.4: testing begins
	add transproxy and masquerading support
	fix syntax problems with iptables and ipchains output
	simple logging support (logs only initiating packets)
	rule checking -- ports only allowed with tcp or udp
	initial icmp support

0.3:
	add ipfilter backend
	fix yet another silly bug in main()
	add warning notice to cisco driver
	cisco driver supports port ranges now

0.2: Various improvements
	rewrote iptables driver
	added ipchains driver
	added beginnings of Cisco IOS ACL driver
		(see TODO for why this doesn't work yet)
	added TODO and HONESTY files

0.1: Initial release
	iptables driver
	a few bugs


pfc.c
/*
 * filter compilation front-end
 *
 * $Id: filtergen.c,v 1.17 2002/11/11 20:51:38 matthew Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <stdarg.h>

#include "filter.h"

#ifndef OLD_FILTERGEN
#include "ast.h"
int yyparse(void *);
void yyrestart(FILE *);
extern struct filter * convert(struct ast_s * n);
#else

struct filter * yyparse(void);
#endif

static FILE *outfile;

int oputs(const char *s)
{
	int r = 0;
	if(s) {
		r = fputs(s, outfile);
		if(r > 0) {
			fputc('\n', outfile);
			r++;
		}
	}
	return r;
}

int oprintf(const char *fmt, ...)
{
	va_list args;
	va_start(args, fmt);
	return vfprintf(outfile, fmt, args);
}


struct filtyp {
	char *name;
	filtergen *compiler;
	filter_flush *flusher;
} filter_types[] = {
	{ "iptables", fg_iptables, flush_iptables, },
	{ "ipchains", fg_ipchains, flush_ipchains, },
	{ "ipfilter", fg_ipfilter, },
	{ "cisco", fg_cisco, },
	{ NULL, },
};


int main(int argc, char **argv)
{
	struct filter *f;
	int l;
	time_t t;
	char buf[100];
	char *filepol = NULL, *ftn = NULL, *ofn = NULL;
	struct filtyp *ft;
	int flags = 0;
	char *progname;
	int arg;

	progname = argv[0];

	while((arg = getopt(argc, argv, "nlmrho:t:F")) > 0) {
		switch(arg) {
		case 'n': flags |= FF_NOSKEL; break;
		case 'l': flags |= FF_LSTATE; break;
		case 'm': flags |= FF_LOOKUP; break;
		case 'h': flags |= FF_LOCAL; break;
		case 'r': flags |= FF_ROUTE; break;
		case 'o': ofn = strdup(optarg); break;
		case 't': ftn = strdup(optarg); break;
		case 'F': flags |= FF_FLUSH; break;
		default: return 1;
		}
	}

	if((flags & FF_LOCAL) && (flags & FF_ROUTE)) {
		fprintf(stderr, "the -h and -r options are mutually exclusive\n");
		return 1;
	}

	if (ofn) {
		/* XXX - open a different tempfile, and rename on success */
		outfile = fopen(ofn, "w");
		if(!outfile) {
			fprintf(stderr, "can't open output file \"%s\"\n", ofn);
			return 1;
		}
	} else
		outfile = stdout;

	if(!ftn || !*ftn) ftn = "iptables";
	for(ft = filter_types; ft->name; ft++)
		if(!strcmp(ftn, ft->name))
			break;
	if(!ft->name) {
		fprintf(stderr, "%s: unknown filter type \"%s\"\n", progname, ftn);
		return 1;
	}

	/* Extra arg (either filename or flush policy) */
	if((filepol = argv[optind])) {
		/* Extra args? */
		while(argv[++optind])
			fprintf(stderr, "%s: extra argument \"%s\" ignored\n",
					progname, argv[optind]);
	}

	/* What to do, then? */
	if(flags & FF_FLUSH) {
		/* Just flush it */
		enum filtertype flushpol;

		if(!filepol || !strcasecmp(filepol, "accept"))
			flushpol = T_ACCEPT;
		else if(!strcasecmp(filepol, "drop"))
			flushpol = DROP;
		else if(!strcasecmp(filepol, "reject"))
			flushpol = T_REJECT;
		else {
			fprintf(stderr, "invalid flush policy \"%s\"\n", filepol);
			return 1;
		}

		l = ft->flusher(flushpol, flags);

	} else {
		/* Compile from a file */
		if(filepol && !strcmp(filepol, "-")) filepol = NULL;

		if(filter_fopen(filepol)) return 1;

#ifdef OLD_FILTERGEN
		f = filter_parse_list();

		if (!f) {
			fprintf(stderr, "couldn't parse file\n");
			return 1;
		}
#else
		{
		    struct ast_s ast;

		    if (yyparse((void *) &ast) == 0) {
			f = convert(&ast);
			if (!f) {
			    fprintf(stderr, "couldn't convert file\n");
			    return 1;
			}
		    } else {
			fprintf(stderr, "couldn't parse file\n");
			return 1;
		    }
		}
#endif

		strftime(buf, sizeof(buf)-1, "%a %b %e %H:%M:%S %Z %Y",
				localtime((time(&t),&t)));
		oprintf("# filter generated from %s via %s backend at %s\n",
			filepol ?: "standard input", ft->name, buf);
		l = ft->compiler(f, flags);
	}

	if(ofn) fclose(outfile);

	if(l < 0) {
		fprintf(stderr, "an error occurred: most likely the filters defined make no sense\n");
		if(ofn) unlink(ofn);
		return 1;
	}
	fprintf(stderr, "generated %d rules\n", l);
	return 0;
}
